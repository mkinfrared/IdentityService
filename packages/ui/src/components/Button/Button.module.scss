@import "../../styles/common";

.Button {
  border: none;
  border-radius: 20px;
  color: #000000;
  cursor: pointer;
  font-weight: 600;
  min-width: 100px;
  padding: 10px 15px;
  position: relative;

  &.opaque {
    background: var(--purple-gradient);
    color: var(--white);
  }

  &.transparent {
    background-color: transparent;
    border: 2px solid var(--purple);
    color: var(--white);
  }

  &.loading {
    color: transparent;

    > * {
      opacity: 0;
      transform: scale(0);
    }

    svg {
      opacity: 1;
      transform: translate(-50%, -50%) scale(1);
    }
  }

  &:disabled {
    cursor: auto;
    filter: grayscale(0.9);
  }

  &:focus-visible {
    @include focusVisible;
  }

  &,
  > * {
    transition: all 100ms ease-in;
  }

  > svg {
    height: calc(16px * 1.8);
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%) scale(0);
  }
}

.Story {

  > div {

    &:not(:last-child) {
      margin-bottom: 20px;
    }
  }
}
