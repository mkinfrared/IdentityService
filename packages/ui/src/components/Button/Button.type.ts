type ButtonAttributes = React.ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = ButtonAttributes & {
  children?: React.ReactNode;
  className?: string;
  loading?: boolean;
  variant?: "opaque" | "transparent";
};
