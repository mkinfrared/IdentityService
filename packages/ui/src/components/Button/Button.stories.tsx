import { Meta, Story } from "@storybook/react";
import React from "react";

import Text from "components/Text";

import { Button } from "./Button";
import css from "./Button.module.scss";
import { ButtonProps } from "./Button.type";

export default {
  title: "UI/Button",
  component: Button,
  argTypes: {
    backgroundColor: { control: "color" },
  },
} as Meta;

const Template: Story<ButtonProps> = (args) => (
  <div className={css.Story}>
    <div>
      <Button {...args}>I am a button</Button>
    </div>
    <div>
      <Button {...args}>
        <Text>I am a button with paragraph</Text>
      </Button>
    </div>
  </div>
);

const Default = Template.bind({});

Default.args = {
  variant: "opaque",
};

export { Default };
