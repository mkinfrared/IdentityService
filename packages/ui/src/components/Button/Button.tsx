/* eslint-disable react/button-has-type */
import { classNames } from "@identity-service/core";
import React, { memo } from "react";

import Spinner from "components/Spinner";

import css from "./Button.module.scss";
import { ButtonProps } from "./Button.type";

const Button = ({
  children,
  className,
  loading = false,
  type = "button",
  variant = "opaque",
  ...rest
}: ButtonProps) => {
  const classes = [css.Button, className];

  switch (variant) {
    case "transparent":
      classes.push(css.transparent);

      break;

    default:
      classes.push(css.opaque);
  }

  if (loading) {
    classes.push(css.loading);
  }

  return (
    <button
      className={classNames(...classes)}
      data-testid="Button"
      type={type}
      {...rest}
    >
      <Spinner type="bars" />
      {children}
    </button>
  );
};

export { Button };

export default memo(Button);
