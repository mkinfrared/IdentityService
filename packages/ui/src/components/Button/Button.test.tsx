import { render } from "@testing-library/react";
import React from "react";

import { Button } from "./Button";

describe("<Button />", () => {
  const Component = <Button />;

  it("should be defined", () => {
    expect(Button).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should render an opaque button by default", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Button");

    expect(element).toBeDefined();

    expect(element).toHaveClass("opaque");
  });

  it("should render a transparent button when variant prop is transparent", () => {
    const { getByTestId } = render(<Button variant="transparent" />);
    const element = getByTestId("Button");

    expect(element).toBeDefined();

    expect(element).toHaveClass("transparent");
  });

  it("should render a disabled button when disabled prop is true", () => {
    const { getByTestId } = render(<Button disabled />);
    const element = getByTestId("Button");

    expect(element).toBeDefined();

    expect(element).toBeDisabled();
  });

  it("should add class name passed in props", () => {
    const classname = "marklar";
    const { getByTestId } = render(<Button className={classname} />);
    const element = getByTestId("Button");

    expect(element).toBeDefined();

    expect(element).toHaveClass(classname);
  });

  it("should add 'loading' class name to button", () => {
    const classname = "loading";
    const { getByTestId } = render(<Button className={classname} loading />);
    const element = getByTestId("Button");

    expect(element).toBeDefined();

    expect(element).toHaveClass(classname);
  });
});
