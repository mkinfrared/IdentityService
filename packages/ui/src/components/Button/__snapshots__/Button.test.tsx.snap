// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match the snapshot 1`] = `
<div>
  <button
    class="Button opaque"
    data-testid="Button"
    type="button"
  >
    <svg
      class="Spinner"
    >
      bars.svg
    </svg>
  </button>
</div>
`;
